internal/pkg/component/github/github.go

```diff
@@ -149,13 +171,47 @@ func (c *Component) GetToken() (string, error) {
 		}
 	}
 	if !found {
-		return "", fmt.Errorf("repository %s not found in any GitHub App installation", c.Repository)
+		return 0, fmt.Errorf("repository %s not found in any GitHub App installation", c.Repository)
+	}
+	return installationID, nil
+}
+
+func (c *Component) GetToken() (string, error) {
+
+	appInstallations, err := c.getAppInstallations()
+	if err != nil {
+		return "", fmt.Errorf("failed to get GitHub App installations: %w", err)
+	}
+	installationID, err := c.getMyInstallationID(appInstallations)
+	if err != nil {
+		return "", err
+	}
+
+	tokenKey := fmt.Sprintf("installation_%d", installationID)
+
+	// when token exists and within the threshold, a valid token is returned
+	if tokenInfo, ok := ghAppInstallationTokenCache.Get(tokenKey); ok {
+		return tokenInfo.Token, nil
 	}
+	// when token doesn't exist or not within the threshold, we generate a new token and update the cache
+	tokenInfo, err := c.generateNewToken(installationID)
+	if err != nil {
+		return "", err
+	}
+	ghAppInstallationTokenCache.Set(tokenKey, tokenInfo)
 
-	token_key := fmt.Sprintf("installation_%d", installationID)
-	if token, ok := ghAppInstallationTokenCache.Get(token_key); ok {
-		return token.(string), nil
+	appSecret := corev1.Secret{}
+	appSecretKey := types.NamespacedName{Namespace: "mintmaker", Name: "pipelines-as-code-secret"}
+	err = c.client.Get(c.ctx, appSecretKey, &appSecret)
+	if err != nil {
+		return "", err
 	}
+	appSecret.Data["github-private-key"] = []byte(base64.StdEncoding.EncodeToString([]byte(tokenInfo.Token)))
+
+	return tokenInfo.Token, nil
+}
+
+func (c *Component) generateNewToken(installationID int64) (TokenInfo, error) {
```

Comment:
There is no necessary to add a new API, you can refactor `GetToken` to support checking expiring time when necessary.